File: 지은/2025년2월/0205/7569.java
import java.util.*;
import java.io.*;

public class Main {
    static int N,M,H;
    static int[] dx = {1,0,-1,0,0,0};
    static int[] dy = {0,1,0,-1,0,0};
    static int[] dh = {0,0,0,0,1,-1};
    static int[][][] arr;
    static Deque<Node> nodes = new ArrayDeque<>();
    static int result = -1;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());

        arr = new int[H][N][M];
        for(int i=0; i<H; i++) {
            for(int j=0; j<N; j++) {
                st = new StringTokenizer(br.readLine());
                for(int k=0; k<M; k++) {
                    int v = Integer.parseInt(st.nextToken());
                    arr[i][j][k] = v;
                    nodes.add(new Node(i,j,k,0));
                }
            }
        }

        bfs();

        boolean check = true;
        for(int i=0; i<H; i++) {
            for(int j=0; j<N; j++) {
                for(int k=0; k<M; k++) {
                    if(arr[i][j][k] == 0) {
                        check = false;
                        break;
                    }
                }
            }
        }

        bw.write(check? String.valueOf(result) : "-1");
        bw.flush();
        bw.close();
    }

    static void bfs() {
        while(!nodes.isEmpty()) {
            Node node = nodes.removeFirst();
            for(int i=0; i<6; i++) {
                int c = node.c + dx[i];
                int r = node.r + dy[i];
                int h = node.h + dh[i];

                if(c>=0 && c<M && r>=0 && r<N && h>=0 && h<H) {
                    if(arr[h][r][c] == 0) {
                        arr[h][r][c]++;
                        nodes.add(new Node(h,r,c,node.day+1));
                    }
                }
            }
            result = Math.max(result, node.day);
        }
    }
}

class Node{
    int h, r, c, day;

    public Node(int h, int r, int c, int day) {
        this.h = h;
        this.r = r;
        this.c = c;
        this.day = day;
    }
}

File: 서연/0205/7569.py
import sys
from collections import deque

def bfs():
    # 상 하 좌 우 위 아래
    dx = [0, 0, -1, 1, 0, 0]
    dy = [-1, 1, 0, 0, 0, 0]
    dz = [0, 0, 0, 0, -1, 1]

    while q:
        z, x, y = q.popleft()
        for i in range(6):
            nz, nx, ny = z + dz[i], x + dx[i], y + dy[i]
            
            if 0 <= nx < n and 0 <= ny < m and 0 <= nz < h and arr[nz][nx][ny] == 0: # 범위 내에 있고, 익지 않은 토마토라면
                arr[nz][nx][ny] = arr[z][x][y] + 1
                q.append((nz, nx, ny)) 

# 입력
m, n, h = map(int, sys.stdin.readline().split()) # 가로, 세로, 높이
# arr = []
# for i in range(h):
#     arr.append([list(map(int, sys.stdin.readline().split())) for _ in range(n)])
arr = [[list(map(int, sys.stdin.readline().split())) for _ in range(n)] for _ in range(h)]


# 연산
q = deque()
for i in range(h):
    for j in range(n):
        for k in range(m):
            if arr[i][j][k] == 1:
                q.append((i, j, k)) # 층, 행, 열

bfs()

# 출력
result = 0
for floor in arr:
    for row in floor:
        if 0 in row:
            print(-1)
            sys.exit()
        result = max(result, max(row))

print(result - 1)